[kaajinn] iP

kaajinn opened the pull request:
# Applemazer
> "His micro adjustments are crazy!" – Anonymous

Applemazer frees your mind of having to remember things you need to do. It's,
- text-based
- easy to learn
- ~~FAST~~ SUPER FAST to use

All you need to do is,
1. download it from [here](https://github.com/kaajinn/ip).
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you 😉

And it is **FREE**!

Features:

- [x] Managing tasks
- [ ] Managing deadlines (coming soon)
- [ ] Reminders (coming soon)
#
If you are a Java programmer, you can use it to practice Java too. Here's the `main` method:

```Java
public class Applemazer {
    public static void main(String[] args) {
        new Applemazer("./data/Applemazer.ser").process();
    }
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

kaajinn committed:
Renamed main class and added greeting/farewell

kaajinn committed:
Added support for echo and removed horizontal spacers.
Reformatted greeting and farewell into functions.

kaajinn committed:
Reverted initial changes to .gitignore

kaajinn committed:
Added support for add/list

kaajinn committed:
Added support for marking/unmarking function
Added Task class

kaajinn committed:
Fixed list command not displaying marked/unmarked status

kaajinn committed:
Added support for tracking three types of tasks (ToDos, Events, Deadlines)

kaajinn committed:
Updated script and test cases used for automated text UI testing

kaajinn committed:
Added error handling for common input errors

kaajinn committed:
Added support for deleting tasks from the list, and added enums

kaajinn committed:
Added support for saving/loading chatbot state

kaajinn committed:
Test commit

kaajinn committed:
Merge branch 'master' into branch-Level-7

kaajinn committed:
Merge branch 'branch-Level-7'

kaajinn committed:
Added support for dates and times
Added DateHandler class

kaajinn committed:
Merge branch 'branch-Level-8'

kaajinn committed:
Refactored all logic to use more OOP

kaajinn committed:
Organized refactored files into packages

kaajinn committed:
Merge branch 'add-gradle-support'

kaajinn committed:
Added gradle support

kaajinn committed:
Added JUnit tests for Parser and DateHandler class

kaajinn committed:
Added JavaDoc comments for methods

kaajinn committed:
Added JavaDocs for classes and modified previous JavaDocs

kaajinn committed:
Improved style of import statements

kaajinn committed:
Added support for find command

kaajinn committed:
Merge branch 'branch-A-CodingStandard'

# Conflicts:
#	src/main/java/Applemazer/Parser.java
#	src/main/java/Applemazer/UI.java
#	src/main/java/Commands/ByeCommand.java
#	src/main/java/Commands/Command.java
#	src/main/java/Commands/DeadlineCommand.java
#	src/main/java/Commands/EventCommand.java
#	src/main/java/Commands/IntegerCommand.java
#	src/main/java/Commands/ListCommand.java
#	src/main/java/Commands/NonExistentCommand.java
#	src/main/java/Commands/TodoCommand.java

kaajinn committed:
Merge branch 'branch-Level-9'

# Conflicts:
#	src/main/java/Tasks/Task.java

kaajinn committed:
Fixed bug for find command

kaajinn committed:
Updated casing for certain file names

kaajinn committed:
Delete src/main/java/Applemazer directory

kaajinn committed:
Delete src/main/java/Commands directory

kaajinn committed:
Delete src/main/java/Tasks directory

kaajinn committed:
Merge branch 'master' of https://github.com/kaajinn/ip

kaajinn committed:
Fix several coding style issues

kaajinn committed:
Use Checkstyle to detect coding style violations

kaajinn committed:
Revert "Use Checkstyle to detect coding style violations"

This reverts commit e4b91624ed7a25e999fcea1a22b8fb71aa3fed54.

kaajinn committed:
Use Checkstyle to detect coding style violations

kaajinn committed:
Merge branch 'branch-A-CheckStyle'

verakohh commented:
### Issue 1: Invalid Date Handling for Events

I noticed that the current implementation allows the creation of events with invalid dates, such as when the 'to' date is earlier than the 'from' date. For example:

``` 
event meeting /from 2022-12-12 /to 2021-12-12
Got it. I've added this task: 
    [E][ ] meeting (from: 12 Dec 2022 to: 12 Dec 2021) 
Now you have 5 tasks in the list. 

Saving data...
Save successful.
```

This could lead to confusion and potentially incorrect data being stored. I suggest adding validation to check that the 'to' date is not earlier than the 'from' date before allowing the task to be created.

### Issue 2: Missing Hi/Hello Command

It looks like the application might be missing a response for a `hi` or `hello` command. Since users might naturally start an interaction with a greeting, adding this command could improve user experience. For instance, a simple response like "Hi there! How can I assist you today?" would be great.

---

### Additional Feedback:
- I really like how the time component is optional for deadlines and events!

verakohh commented on src/main/java/applemazer/Parser.java:
```diff
@@ -0,0 +1,177 @@
+        int taskNumber;
+
+        return switch (command) {
+        case "bye" -> {
```
Missing Hi/Hello Command

It looks like the application might be missing a response for a `hi` or `hello` command. Since users might naturally start an interaction with a greeting, adding this command could improve user experience. For instance, a simple response like "Hi there! How can I assist you today?" would be great.

verakohh commented on src/main/java/commands/EventCommand.java:
```diff
@@ -0,0 +1,62 @@
+            tasks.add(task);
+            task.printTaskAddedMessage(tasks.size());
+            storage.save();
+        } catch (DateTimeException e) {
```
I noticed that the current implementation allows the creation of events with invalid dates, such as when the 'to' date is earlier than the 'from' date. For example:

``` 
event meeting /from 2022-12-12 /to 2021-12-12
Got it. I've added this task: 
    [E][ ] meeting (from: 12 Dec 2022 to: 12 Dec 2021) 
Now you have 5 tasks in the list. 

Saving data...
Save successful.
```

This could lead to confusion and potentially incorrect data being stored. I suggest adding validation to check that the 'to' date is not earlier than the 'from' date before allowing the task to be created.

verakohh reviewed (commented):
I really like how the time component is optional for deadlines and events!

kaajinn committed:
Set up initial GUI implementation and remove text-based UI

kaajinn committed:
Minor tweaks to GUI and added timer before shutdown

kaajinn committed:
Merge branch 'branch-Level-10'

kaajinn committed:
Remove unintended println

ashea10 commented on src/main/java/applemazer/Applemazer.java:
```diff
@@ -0,0 +1,83 @@
+    private static Storage storage;
+    private final Parser parser; // For text-based UI.
+    private final Ui ui;
+    private boolean isProcessing = true;
```
Good job on being consistent with the naming standard for booleans, noticed these elsewhere as well!

ashea10 commented on src/main/java/applemazer/Applemazer.java:
```diff
@@ -0,0 +1,83 @@
+        }
+    }
+
+    public Ui getUi() {
+        return this.ui;
+    }
+
+    public boolean shouldExit() {
+        return !this.isProcessing;
+    }
+
```
Missing JavaDoc comments, would be great if you could add it in for all methods/classes.

ashea10 commented on src/main/java/applemazer/Parser.java:
```diff
@@ -0,0 +1,177 @@
+        };
+    }
+
+    private void parseByeCommand() throws Exception {
+        String desc = sc.nextLine().trim();
+        if (!desc.isEmpty()) {
+            throw new Exception("OOPS!!! The description of bye should be empty. ");
+        }
+    }
```
Would have been great if you have made use of custom exception(s).

ashea10 commented on src/main/java/applemazer/Storage.java:
```diff
@@ -0,0 +1,94 @@
+     * <p>
+     * If the stored filepath does not exist, a new task list will be returned.
+     */
+    @SuppressWarnings("unchecked")
```
To avoid the use of `SuppressWarnings` here, maybe you could use a method to convert the format read from the file to the format of a "task", so that we can do something like `Task.toProperFormat(str.readObject)`.

ashea10 commented on src/main/java/applemazer/Ui.java:
```diff
@@ -0,0 +1,85 @@
+    /**
+     * Constructor for the Ui object.
+     */
+    public Ui() {}
```
Empty constructor is present by default in Java. If it is the only constructor, it isn't required, though not wrong if it is there as well.

ashea10 commented on src/main/java/commands/Command.java:
```diff
@@ -0,0 +1,37 @@
+    /**
+     * Represents commands that handle integer input.
+     */
+    public enum IntegerCommands {
```
Good job on using Enums!

ashea10 commented on src/main/java/tasks/Task.java:
```diff
@@ -0,0 +1,56 @@
+     * Returns the task description.
+     * @return The task description.
+     */
+    @Override
```
Both methods are doing the same thing it seems, perhaps you can remove one and just stick with toString.

ashea10 reviewed (commented):
LGTM, good job! Just a few nits on the design choices.

kaajinn committed:
Enable assertions in build.gradle

kaajinn committed:
Add assertions before certain operations to ensure correctness

kaajinn committed:
Improve code quality and code standard

kaajinn committed:
Merge pull request #1 from kaajinn/branch-A-Assertions

Add assertions before certain operations to ensure correctness

kaajinn committed:
Merge branch 'master' of https://github.com/kaajinn/ip

kaajinn committed:
Merge branch 'master' into branch-A-CodeQuality

# Conflicts:
#	src/main/java/applemazer/Applemazer.java
#	src/main/java/applemazer/Ui.java

kaajinn committed:
Merge pull request #2 from kaajinn/branch-A-CodeQuality

Improve code quality and code standard

kaajinn committed:
Merge branch 'master' of https://github.com/kaajinn/ip

kaajinn committed:
Task list no longer allows user to add duplicate tasks

Task list allows user to add duplicate tasks.

Allowing users to add duplicate tasks introduces clutter into the
list and makes it harder for the user to read.

Let's store the task descriptions of newly added tasks into a map
so that we can detect when a duplicate task (i.e. same description
and/or date) has been added.

Using a map is preferable as it has O(1) lookup time.

kaajinn committed:
Add Javadoc comments for methods related to duplicate task
handling

There are no Javadoc comments for methods related to duplicate task
handling.

Adding Javadoc comments allows other developers to understand the
purpose of the DuplicateHandler class and its related methods.

Let's
* use the Javadoc feature in IntelliJ to help structure the Javadoc
comment skeleton
* fill the Javadoc comment skeleton with the relevant information

kaajinn committed:
Merge branch 'branch-BCD-Extension'

kaajinn committed:
Improve coding standard

kaajinn committed:
Improve coding standard

kaajinn committed:
Set title name for JavaFX window

kaajinn committed:
Add representative screenshot

kaajinn committed:
Update README.md

kaajinn committed:
Update README.md

kaajinn committed:
Update README.md

kaajinn committed:
Merge branch 'master' of https://github.com/kaajinn/ip

kaajinn committed:
Improve coding standard

kaajinn committed:
Add more tests for JUnit testing

kaajinn committed:
Add resizable GUI

kaajinn committed:
Update Ui.png

kaajinn committed:
Remove original Duke README.md file

kaajinn committed:
Remove all saved entries from .ser file
